fortex/health/processors/coreference_processor.py

```diff
@@ -0,0 +1,209 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Coreference Processor
+"""
+from typing import Dict, Set
+import importlib
+
+import neuralcoref
+
+from forte.common import Resources, ProcessExecutionException
+from forte.common.configuration import Config
+from forte.data.data_pack import DataPack
+from forte.processors.base import PackProcessor
+
+from ft.onto.base_ontology import CoreferenceGroup
+from ftx.medical.clinical_ontology import MedicalArticle
+
+from fortex.spacy.spacy_processors import load_lang_model
+
+__all__ = [
+    "CoreferenceProcessor",
+]
+
+
+class CoreferenceProcessor(PackProcessor):
+    r"""
+    Implementation of this CoreferenceProcessor has been based on huggingface
+    NeuralCoref. You can find more details in the original repo.
+
+    Note that the NeuralCoref package from PyPI uses a dated spaCy
+    version (2.1), which can cause segmentation fault with the spaCy we use (2.3).
+    Please install NeuralCoref by building from source.
+
+    Referred repository link:
+    https://github.com/huggingface/neuralcoref
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.spacy_nlp = None
+
+    def set_up(self, configs: Config):
+        self.spacy_nlp = load_lang_model(configs.lang)
+
+        if self.spacy_nlp is None:
+            raise ProcessExecutionException(
+                "The SpaCy pipeline is not initialized, maybe you "
+                "haven't called the initialization function."
+            )
+
+        model = configs.model
+        cfg_inference = {
+            "greedyness": configs.greedyness,
+            "max_dist": configs.max_dist,
+            "max_dist_match": configs.max_dist_match,
+            "blacklist": configs.blacklist,
+            "store_scores": configs.store_scores,
+            "conv_dict": configs.conv_dict,
+        }
+        neuralcoref.add_to_pipe(
+            self.spacy_nlp, model=model, cfg_inference=cfg_inference
+        )
+
+    def initialize(self, resources: Resources, configs: Config):
+        super().initialize(resources, configs)
+        self.set_up(configs)
+
+    def _process(self, input_pack: DataPack):
+        r"""
+        Coreference resolution is done by
+        a spaCy pipeline with `NeuralCoref` added.
+
+        Then we translate the output to `CoreferenceGroup` and
+        `MedicalEntityMention`
+        """
+
+        def load_module(string):
+            path_str, module_str = string.rsplit(".", 1)
+            mod = importlib.import_module(path_str)
+            return getattr(mod, module_str)
+
+        # Default: Document
+        entry_type = load_module(self.configs.entry_type)
+
+        # Default: MedicalEntityMention
+        mention_type = load_module(self.configs.mention_type)
+
+        for entry_specified in input_pack.get(entry_type=entry_type):
+            result = self.spacy_nlp(entry_specified.text)
+
+            article = MedicalArticle(
```

Comment:
Is there a reason why we are using `MedicalArticle` here instead of `CoreferenceGroup`? `MedicalArticle` is a different kind of annotation. `CoreferenceGroup` is more suited to this use case I feel.

Reply:
I saw `CoreferenceGroup` is used below. But also attached to `MedicalArticle`. You don't need to do so. If you run `article.get(CoreferenceGroup)`, you will get all the groups in this article (calculated by Forte using mention spans.)

Reply:
Got it.

Reply:
Removed `MedicalArticle`.